common/membership/hashring_test.go

```diff
@@ -273,27 +268,41 @@ func TestFailedRefreshLogsError(t *testing.T) {
 	}).Times(1)
 
 	wg.Add(1) // another call to GetMembers should happen because of handleUpdate
-	td.bypassRefreshRatelimiter()
 	td.hashRing.handleUpdates(ChangedEvent{})
 
 	require.True(t, common.AwaitWaitGroup(&wg, maxTestDuration), "GetMembers must be called again (and fail)")
 	td.hashRing.Stop()
 	assert.Equal(t, 1, td.observedLogs.FilterMessageSnippet("failed to refresh ring").Len())
 }
 
-func TestRefreshUpdatesRingOnlyWhenRingHasChanged(t *testing.T) {
+func TestRefreshUpdatesSubscribersOnlyWhenMembershipHasChanged(t *testing.T) {
+	membership := randomHostInfo(3)
 	td := newHashringTestData(t)
+	changeCh := make(chan *ChangedEvent, 2)
 
-	td.mockPeerProvider.EXPECT().Subscribe(gomock.Any(), gomock.Any()).Times(1)
-	td.mockPeerProvider.EXPECT().GetMembers("test-service").Times(1).Return(randomHostInfo(3), nil)
+	var updateHandler func(event ChangedEvent)
+	td.mockPeerProvider.EXPECT().Subscribe("test-service", gomock.Any()).Do(
+		func(name string, fn func(event ChangedEvent)) {
+			updateHandler = fn
+		}).Times(1)
+
+	td.mockPeerProvider.EXPECT().GetMembers("test-service").Return(membership, nil).Times(2)
 	td.mockPeerProvider.EXPECT().WhoAmI().AnyTimes()
 
-	// Start will also call .refresh()
+	require.NoError(t, td.hashRing.Subscribe("subscriber1", changeCh))
 	td.startHashRing()
-	updatedAt := td.hashRing.members.refreshed
 
-	assert.NoError(t, td.hashRing.refresh())
-	assert.Equal(t, updatedAt, td.hashRing.members.refreshed)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		<-changeCh
+	}()
+	assert.True(t, common.AwaitWaitGroup(&wg, maxTestDuration), "first notification has to be there")
```

Comment:
maybe worth mentioning that this is the "ring-started" refresh notification, as "first" is a bit vague